File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
int res;
public:
    int diameterOfBinaryTree(TreeNode* root) {
        res = 0;
        if (!root) return res;
        maxHeight(root);

        return res;
        
    }

    int maxHeight(TreeNode* node) {

        if (!node) return 0;

        int left = maxHeight(node -> left);
        int right = maxHeight(node -> right);

        res = max(res, left + right);

        return 1 + max(left, right);

    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.res = 0
        self.maxHeight(root)
        
        return self.res
    
    def maxHeight(self, root) -> int:
        if not root:
            return 0
    
        left = self.maxHeight(root.left)
        right = self.maxHeight(root.right)

        self.res = max(self.res, left + right)

        return 1 + max(left, right)
        